sensors/imu/mpu6000.c

```diff
@@ -0,0 +1,317 @@
+} mpu6000_ctx_t;
+
+
+static int16_t translateGyr(uint8_t hbyte, uint8_t lbyte)
+{
+	int16_t val = 0;
+	val |= hbyte;
+	val <<= 8;
+	val |= lbyte;
+
+	/* overflow handling */
+	if (val > GYR_OVERFLOW) {
+		return INT16_MAX;
+	}
+	if (val < -GYR_OVERFLOW) {
+		return INT16_MIN;
+	}
+	return GYR2000DPS_CONV_MRAD * val; /* sensor value to [mrad/s] */
+}
+
+
+static int16_t translateAcc(uint8_t hbyte, uint8_t lbyte)
+{
+	int16_t val = 0;
+	val |= hbyte;
+	val <<= 8;
+	val |= lbyte;
+	return ACC8G_CONV_MG * MG_CONV_MMS2 * val; /* sensor value to [mm/s^2] */
+}
+
+
+static int spiWriteReg(mpu6000_ctx_t *ctx, uint8_t regAddr, uint8_t regVal)
+{
+	unsigned char cmd[2] = { (regAddr & 0x7F), regVal }; /* write bit set to regAddr */
+
+	return sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, cmd, sizeof(cmd), NULL, 0, sizeof(cmd));
+}
+
+
+static int mpu6000_whoamiCheck(mpu6000_ctx_t *ctx)
+{
+	uint8_t cmd, val;
+	int err;
+
+	cmd = REG_WHOAMI | SPI_READ_BIT;
+	val = 0;
+	err = sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), &val, sizeof(val), sizeof(cmd));
+	if ((err < 0) || (val != VAL_WHOAMI)) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int mpu6000_hwSetup(mpu6000_ctx_t *ctx)
+{
+	if (mpu6000_whoamiCheck(ctx) != 0) {
+		printf("mpu6000: cannot read/wrong WHOAMI returned!\n");
+		return -1;
+	}
+
+	/* Reset procedure according to "MPU-6000/MPU-6050 Register Map and Descriptions", section 4.28 */
+	if (spiWriteReg(ctx, REG_PWR_MGMT_1, VAL_PWR_MGMT_1_DEVICE_RESET) < 0) {
+		return -1;
+	}
+	usleep(100 * 1000);
+	if (spiWriteReg(ctx, REG_SIGNAL_PATH_RESET, VAL_SIGNAL_PATH_RESET_ACCEL_RESET | VAL_SIGNAL_PATH_RESET_GYRO_RESET | VAL_SIGNAL_PATH_RESET_TEMP_RESET) < 0) {
+		return -1;
+	}
+	usleep(100 * 1000);
+
+	/* Wake up the device. Writing 0 unsets DEVICE_RESET, SLEEP, CYCLE, TEMP_DIS bits */
+	if (spiWriteReg(ctx, REG_PWR_MGMT_1, (0 | VAL_PWR_MGMT_1_CLKSEL_PLL_Z)) < 0) {
+		return -1;
+	}
+
+	/* Disabling I2C */
+	if (spiWriteReg(ctx, REG_USER_CTRL, VAL_USER_CTRL_I2C_IF_DIS) < 0) {
+		return -1;
+	}
+
+	/* Sampling rate and LPF config */
+	if (spiWriteReg(ctx, REG_CONFIG, VAL_CONFIG_DLPF_CFG_256HZ) < 0) {
+		return -1;
+	}
+
+	if (spiWriteReg(ctx, REG_SMPRT_DIV, VAL_SMPRT_DIV_1) < 0) {
+		return -1;
+	}
+
+	/* Sensor ranges setup */
+	if (spiWriteReg(ctx, REG_GYRO_CONFIG, VAL_GYRO_CONFIG_FS_SEL_2000DPS) < 0) {
+		return -1;
+	}
+
+	if (spiWriteReg(ctx, REG_ACCEL_CONFIG, VAL_ACCEL_CONFIG_AFS_SEL_8G) < 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static void mpu6000_threadPublish(void *data)
+{
+	static uint32_t dAngleX = 0, dAngleY = 0, dAngleZ = 0;
+	static time_t tStampLast;
+
+	int err;
+	time_t tStamp, step;
+	sensor_info_t *info = (sensor_info_t *)data;
+	mpu6000_ctx_t *ctx = info->ctx;
+	uint8_t ibuf[SENSOR_OUTPUT_SIZE] = { 0 };
+	uint8_t obuf;
+
+	gettime(&tStampLast, NULL);
+
+	/* TODO: SPI speed bottleneck: MPU6000 SPI accepts <1MHz CLK writes, but <20MHz CLK reads; runtime SPI CLK setup unavailable */
```

Comment:
I've meant setting the speed as desired to allow seamless transition to the new SPI that handles this feature. But if you find this not needed for now it can stay as is